# Exercise 4.7. Deriving New Variables Part 1 (Practice)

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os 

In [2]:
# Set path
path = r'C:\Users\karya\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\karya\\Instacart Basket Analysis'

In [4]:
# Importing ords_prods_merged
df_ords_prods_merged = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [5]:
df_ords_prods_merged.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merged2'],
      dtype='object')

## If- Statements with User-Defined Functions

In [6]:
# Create a subset of 1m rows
df = df_ords_prods_merged[:1000000]

In [7]:
df.shape

(1000000, 15)

In [8]:
# Define a function

def price_label(row):
    
    if row['prices'] <= 5:
        return 'Low-range product'
    elif (row['prices'] > 5) and (row['prices'] <= 15):
        return 'Mid-range product'
    elif (row['prices'] > 15):
        return 'High range product'
    else: return 'Not enough data'
        

In [9]:
# Apply the function
df['price_range'] = df.apply(price_label, axis = 1)

C:\Users\karya\AppData\Local\Temp\ipykernel_2284\2004007646.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis = 1)


In [10]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

In [11]:
# Max function
df['prices'].max()

14.8

In [12]:
df.columns

Index(['order_id', 'user_id', 'order_number', 'orders_day_of_week',
       'order_time_of_day', 'days_since_prior_order', 'product_id',
       'add_to_cart_order', 'reordered', '_merge', 'product_name', 'aisle_id',
       'department_id', 'prices', '_merged2', 'price_range'],
      dtype='object')

## If-Statements with the loc() Functions

In [12]:
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\karya\AppData\Local\Temp\ipykernel_14228\1169838859.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'
C:\Users\karya\AppData\Local\Temp\ipykernel_14228\1169838859.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
# count the number of products within each label ,using the value_counts() function

df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product    756450
Low-range product    243550
Name: count, dtype: int64

## Repeat the Process ---Using ords_prods_merge instead of df:

In [16]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\karya\AppData\Local\Temp\ipykernel_14228\3564270281.py:1: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise in a future error of pandas. Value 'High-range product' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.
  df_ords_prods_merged.loc[df_ords_prods_merged['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [17]:
df_ords_prods_merged.loc[(df_ords_prods_merged['prices'] <= 15) & (df_ords_prods_merged['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [18]:
df_ords_prods_merged.loc[df_ords_prods_merged['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [19]:
df_ords_prods_merged['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860860
Low-range product     10126321
High-range product      417678
Name: count, dtype: int64

## If-Statements with For-Loops

In [20]:
df_ords_prods_merged['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [21]:
# setting the loop
result = []

for value in df_ords_prods_merged["orders_day_of_week"]:
    if value == 0:
        result.append("Busiest day")
    elif value == 4:
        result.append("Least busy")
    else:
        result.append("Regularly busy")

In [22]:
result

['Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Reg

In [23]:
# Add result to the dataframe
df_ords_prods_merged["Busiest day"] = result

In [24]:
# Check the value counts
df_ords_prods_merged["Busiest day"].value_counts(dropna = False)

Busiest day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 4.7. Deriving New Variabels--Part 2

### Q2. Create a new column for the busiest days

In [25]:
# If-Statements with For-loop

result_var2 = []

for value in df_ords_prods_merged["orders_day_of_week"]:
    if value == 0 or value == 1:
        result_var2.append("Busiest days")
    elif value == 3 or value == 4:
        result_var2.append("Least busy days")
    else:
        result_var2.append("Regularly busy")

In [26]:
# Add result to the dataframe

df_ords_prods_merged["Busiest days"] = result_var2

In [27]:
# Check the columns

df_ords_prods_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_time_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_since_prior_order  product_id  add_to_cart_order  reordered _merge  \
0                     NaN       196.0                1.0        0.0   both   
1                    15.0       196.0                1.0        1.0   both   
2                    21.0       196.0                1.0        1.0   both   
3                    29.0       196.0                1.0        1.0   both   
4                    28.0       196.0                1.0        1.0   both   

  product_name  aisle_id  department_id  prices _merged2    price_range_loc  \
0         Soda        77              7     9.0     both  Mid-range product   
1         Soda        77              7     9.0     both  Mid-range product   
2         Soda        77              7     9.0     both  Mid-range product   
3         Soda        77              7     9.0     both  Mid-range product   
4         Soda        77              7     9.0     both  Mid-range product   

      Busiest day     Busiest days  
0  Regularly busy   Regularly busy  
1  Regularly busy  Least busy days  
2  Regularly busy  Least busy days  
3      Least busy  Least busy days  
4      Least busy  Least busy days

### 3. Check the values of this new column for accuracy. Note any observations in markdown format.

In [28]:
# Check the value counts

df_ords_prods_merged["Busiest days"].value_counts(dropna = False)

Busiest days
Regularly busy     12916111
Busiest days       11864412
Least busy days     7624336
Name: count, dtype: int64

#### When cross-checked with the original frquency counts calculate earlier in this exercise the value count of the " Busiest days"  (11864412) Is equal to the sum of the value counts of value 0 (6204182) and value 1 (5660230). And the value count of the "Least buy days"  (7624336) is the sum of the value counts of value 3 (3840534) and value 4 (3783802)

# Q4. Creating a new column called (busiest_period_of_day)

In [29]:
# First lets Check the frequency of values in the "order_hour_of_day" column 

df_ords_prods_merged['order_time_of_day'].value_counts()

order_time_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

####  The senior technical officer at Instacart wants these to be grouped as "Most orders" = 10, 11,14,15,13,12,16,9 Average orders = 17,8,18,19,20,7,21,22  Fewest orders = 23,6,0,1,5,2,4,3

In [30]:
# If-Statements with For-loop

hour_of_day = []

for value in df_ords_prods_merged["order_time_of_day"]:
    if value in [10, 11, 14, 15, 13, 12, 16, 9]:
        hour_of_day.append("Most orders")
    elif value in [17, 8, 18, 19, 20, 7, 21, 22]:
        hour_of_day.append("Average orders")
    else:
        hour_of_day.append("Fewest orders")

In [31]:
# Add result to the dataframe

df_ords_prods_merged['busiest_period_of_day'] = hour_of_day

In [32]:
# Lets check
df_ords_prods_merged[['order_time_of_day', 'busiest_period_of_day']].head(20)

order_time_of_day busiest_period_of_day
0                   8        Average orders
1                   7        Average orders
2                  12           Most orders
3                   7        Average orders
4                  15           Most orders
5                   7        Average orders
6                   9           Most orders
7                  14           Most orders
8                  16           Most orders
9                   8        Average orders
10                  9           Most orders
11                 16           Most orders
12                 11           Most orders
13                  9           Most orders
14                 10           Most orders
15                 14           Most orders
16                 12           Most orders
17                 17        Average orders
18                 10           Most orders
19                  8        Average orders

### Q5. Print the frequency for this new column.

In [33]:
df_ords_prods_merged['busiest_period_of_day'].value_counts()

busiest_period_of_day
Most orders       21118071
Average orders     9997651
Fewest orders      1289137
Name: count, dtype: int64

### Q6.Ensure your notebook is clean and structured and that your code is well commented.

### Q7.Export your dataframe as a pickle file (since you added new columns) and store it correctly in your “Prepared Data” folder.

In [34]:
# Exporting the new dataframe to the 02 Data in the Project folder, using the pickle formate

df_ords_prods_merged.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_updated.pkl'))